enterprise/server/remote_execution/casfs/casfs.go

```diff
@@ -0,0 +1,653 @@
+		}
+		if n.cfs.verbose {
+			log.Debugf("[%s] Open %q%s: serve from scratch file %q", n.cfs.taskID(), relPath, hash, scratchPath)
+		}
+		fd, err := syscall.Open(scratchPath, int(flags), 0)
+		if err != nil {
+			log.Warningf("[%s] Open %q%s failed: %s", n.cfs.taskID(), scratchPath, hash, err)
+			return nil, 0, syscall.EIO
+		}
+		lf := &instrumentedLoopbackFile{FileHandle: fs.NewLoopbackFile(fd), cfs: n.cfs, fd: fd, fullPath: scratchPath}
+		return lf, 0, 0
+	}
+
+	if n.fileNode != nil {
+		if n.cfs.verbose {
+			log.Debugf("[%s] Open %q: fetch %s from CAS", n.cfs.taskID(), relPath, n.fileNode.GetDigest().GetHash())
+		}
+
+		fileMap := dirtools.FileMap{
+			digest.NewKey(n.fileNode.GetDigest()): {&dirtools.FilePointer{
+				FullPath:     scratchPath,
+				RelativePath: relPath,
+				FileNode:     n.fileNode,
+			}},
+		}
+
+		if err := os.MkdirAll(filepath.Dir(scratchPath), 0755); err != nil {
+			log.Warningf("[%s] Could not make dirs for %q: %s", n.cfs.taskID(), scratchPath, err)
+			return nil, 0, syscall.EINVAL
+		}
+
+		err := n.cfs.fetchFiles(fileMap, &dirtools.DownloadTreeOpts{})
+		if err != nil {
+			log.Warningf("[%s] Open fetch %q failed: %s", n.cfs.taskID(), n.fileNode.GetDigest().GetHash(), err)
+			return nil, 0, syscall.EIO
+		}
+
+		n.scratchFile = scratchPath
+
+		fd, err := syscall.Open(scratchPath, int(flags), 0)
+		if err != nil {
+			log.Warningf("[%s] Open %q failed: %s", n.cfs.taskID(), scratchPath, err)
+			return nil, 0, syscall.EIO
+		}
+		lf := &instrumentedLoopbackFile{FileHandle: fs.NewLoopbackFile(fd), cfs: n.cfs, fd: fd, fullPath: scratchPath}
+		return lf, 0, 0
+	}
+
+	return nil, 0, syscall.ENOTSUP
+}
+
+func (n *Node) Create(ctx context.Context, name string, flags uint32, mode uint32, out *fuse.EntryOut) (node *fs.Inode, fh fs.FileHandle, fuseFlags uint32, errno syscall.Errno) {
+	if n.cfs.verbose {
+		log.Debugf("[%s] Create %q", n.cfs.taskID(), filepath.Join(n.relativePath(), name))
+	}
+
+	scratchPath := filepath.Join(n.scratchPath(), name)
+	if err := os.MkdirAll(filepath.Dir(scratchPath), 0755); err != nil {
+		log.Infof("[%s] Could not make dirs for %q: %s", n.cfs.taskID(), scratchPath, err)
+		return nil, 0, 0, syscall.EINVAL
+	}
+
+	fd, err := syscall.Open(scratchPath, int(flags), mode)
+	if err != nil {
+		log.Warningf("[%s] Could not open %q: %s", n.cfs.taskID(), scratchPath, err)
+		return nil, nil, 0, fs.ToErrno(err)
+	}
+	lf := &instrumentedLoopbackFile{FileHandle: fs.NewLoopbackFile(fd), cfs: n.cfs, fd: fd, fullPath: scratchPath}
+	child := &Node{cfs: n.cfs, parent: n, scratchFile: scratchPath}
+	inode := n.cfs.root.NewPersistentInode(ctx, child, fs.StableAttr{Mode: fuse.S_IFREG})
+	if !n.AddChild(name, inode, false) {
+		log.Warningf("[%s] Could not add child %q to %q", n.cfs.taskID(), name, n.relativePath())
+		return nil, nil, 0, syscall.EIO
+	}
+
+	st := syscall.Stat_t{}
+	if err := syscall.Fstat(fd, &st); err != nil {
+		_ = syscall.Close(fd)
```

Comment:
consider logging this as a warning? might indicate a problem if we're failing to close files